packages/language/src/preprocessor/compiler-options-processor.ts

```diff
@@ -11,44 +11,154 @@
 
 export interface CompilerOptionsProcessorResult {
   result: CompilerOptionResult | undefined;
   text: string;
 }
 
 export class CompilerOptionsProcessor {
-  extractCompilerOptions(text: string): CompilerOptionsProcessorResult {
+  // constant for the *PROCESS token length
+  private static readonly PROCESS_TOKEN_LENGTH = 8;
+
+  /**
+   * Extracts compiler options from the given text
+   * @param text - The source text containing compiler options.
+   * @param uri - The URI of the source file, for configuration lookup.
+   * @returns Processed compiler options and modified text.
+   */
+  extractCompilerOptions(
+    text: string,
+    uri: string,
+  ): CompilerOptionsProcessorResult {
     const range = this.getCompilerOptionsRange(text);
+    let srcCompilerOpts: string | undefined = undefined;
+    let newText = text;
+
     if (range) {
       // Magic number 8 is the length of the string "*PROCESS"
-      const offset = range.start + 8;
-      const optionsText = text.substring(offset, range.end);
-      const newText =
+      const offset =
+        range.start + CompilerOptionsProcessor.PROCESS_TOKEN_LENGTH;
+      srcCompilerOpts = text.substring(offset, range.end);
+      newText =
         text.substring(0, range.start) +
         " ".repeat(range.end - range.start) +
         text.substring(range.end);
-      const abstractOptions = parseAbstractCompilerOptions(optionsText, offset);
-      const compilerOptionsResult = translateCompilerOptions(abstractOptions);
-      // Prepend the *PROCESS token to the compiler options result
-      compilerOptionsResult.tokens.unshift(range.token);
+    }
+
+    // Retrieve compiler options from the plugin configuration provider
+    const programConfig =
+      PluginConfigurationProviderInstance.getProgramConfig(uri);
+    const processGroupConfig = programConfig
+      ? PluginConfigurationProviderInstance.getProcessGroupConfig(
+          programConfig.pgroup,
+        )
+      : undefined;
+
+    let configCompilerOpts: string | undefined = undefined;
+    if (processGroupConfig?.["compiler-options"]?.length) {
+      configCompilerOpts = processGroupConfig["compiler-options"].join(" ");
+    }
+
+    if (srcCompilerOpts || configCompilerOpts) {
+      let configCompilerResult: CompilerOptionResult | undefined = undefined;
+      let srcCompilerResult: CompilerOptionResult | undefined = undefined;
+
+      // process compiler options from the config
+      if (configCompilerOpts) {
+        const configAbstractOptions =
+          parseAbstractCompilerOptions(configCompilerOpts);
+        configCompilerResult = translateCompilerOptions(configAbstractOptions);
+        // adjust all issues from config
+        this.adjustIssuesFromConfig(configCompilerResult.issues, range);
+        if (!range) {
+          // don't report config issues if we don't have a *PROCESS directive to attach them to
+          configCompilerResult.issues = [];
+        }
+      }
```

Comment:
Suggestion: I think we should parse the config outside of the compiler options extraction (that happens on every parse of every compilation unit). This makes more sense conceptually (since the compiler options come from outside of the file anyway) and should also save cycles, as we don't need to reparse the compiler options every time.

We can store the parsed and configured compiler options separately on the compilation unit and then merge them for use in the tokenizer, etc. I.e. have three versions of the compiler options available, config, text and merged. This would also resolve the issue from above (printing the warnings).